### DISCUSSION TITLE
Debugging slow ProviderScope rebuilds

### DISCUSSION THREAD

**Discussion by btrautmann** [rrousselGit/riverpod#3126] [Q&A]
I originally [tweeted](https://twitter.com/BrandoTrautmann/status/1724078161002950678) about an issue I'm seeing where I'm getting jank when changing a data point that causes a provider at the tip of the "provider tree" to be rebuilt. In short, the rebuilding of the ProviderScope takes anywhere from 20-30ms. 

For context, the application I'm building offers a way to visualize financial transaction data. It syncs with a remote server and stores the transactions in a local `Isar` database. Many providers observe the transactions via a `StreamProvider` and perform data manipulation on them. Here's one such "chain" of providers:

```dart
/// Provides the on-budget transactions in the current [DateRange] irrespective of the current view.
final onBudgetTransactionsProvider = StreamProvider<List<Transaction>>((ref) async* {
  // transactionsProvider is a `StreamProvider` listening to query changes on the `Isar` database.
  final transactions = await ref.watch(transactionsProvider.future);
  final accounts = await ref.watch(onBudgetAccountsProvider.future);
  yield transactions.whereForAccounts(accounts).toList();
});
```

```dart
/// Provides the on-budget transactions for the current view in the current [DateRange].
final onBudgetTransactionsInCurrentViewProvider = StreamProvider<List<Transaction>>(
  (ref) async* {
    final transactions = await ref.watch(onBudgetTransactionsProvider.future);
    final categories = await ref.watch(categoriesInCurrentViewProvider.future);
    yield transactions.filterForCategories(categories).toList();
  },
);
```

```dart
/// Provides a Map of [Category]s to [Transaction]s that are considered expenses. This is the
/// most common way of looking at expenses.
final expensesInCurrentViewProvider = StreamProvider<Map<Category, List<Transaction>>>((ref) async* {
  final transactions = await ref.watch(onBudgetTransactionsInCurrentViewProvider.future);
  final categories = await ref.watch(expenseCategoriesProvider.future);
  final expenseFilter = ref.watch(expenseFilterProvider);

  bool categoryFilter(BaseTransaction transaction, Category category) {
    if (!expenseFilter(transaction)) return false;
    return transaction.categoryId == category.id;
  }

  yield categories.fold<Map<Category, List<Transaction>>>(
    {},
    (map, category) {
      final matches = transactions.filter((t) => categoryFilter(t, category));
      return map..[category] = matches;
    },
  );
});
```

I did some rudimentary probing at how long these providers take to compute their data values and they can be anywhere from a few hundred ms to several seconds. From what I can tell, this _should_ be fine given how `StreamProviders` will simply emit a loading state and then later a data state once something is `yield`ed.

That said, I'm realizing that part of my being unable to figure out what's causing the jank here is that I'm not sure what parts of my code impact the `ProviderScope` rebuild time. I'm hoping folks can give me some pointers on where I might be doing something really wrong based on the above snippets.

**Comment by rrousselGit** [comment]
The first step would be to get the list of possible problematic providers


I'd suggest creating a ProviderObserver. Then adding a print in didUpdateProvider

From there, the next step would be to use that information to get the list of all providers that rebuild in your jank frame.
This would tell you which ones are likely responsible

**Reply by rrousselGit** [reply]
Note that using StreamProvider is useless in all of the examples you gave.
They could all be replaced with a FutureProvider where by replacing "yield" with "future"

**Reply by btrautmann** [reply]
So I did make _some_ progress. After giving every `Provider` a name and adding a `ProviderObserver` that logged additions/disposals/updates, I noticed that the `Provider` that provided the transaction data was being disposed twice when changing the date range. This was because I stored the `from` date and `to` date in two different key/value pairs in shared preferences, and updating both would cause 2 stream emissions (using RxDart's `CombineStream`). I changed this to have the dates stored in the same key (since they'll always be changed together), and this results in a single emission. **Effectively, this means that any `Provider`s that depend on the transactions `Provider` now rebuild only once instead of twice.** 

However, the `ProviderScope` rebuild time remained the same. This makes sense to me, because it doesn't seem like the time a `FutureProvider` or `StreamProvider` takes to be built should impact the time `ProviderScope` takes to rebuild since they are `async`. Is that a correct assumption? When you say:

> From there, the next step would be to use that information to get the list of all providers that rebuild in your jank frame.

Let's say 25 `Provider`s rebuild when the date range is changed. I could use a `Stopwatch` to measure rough timings on each of their build durations, but some of them are 1+ seconds. This doesn't tell me much because obviously the `ProviderScope` is not waiting a full second to be considered "built", so what, other than duration itself, should I be looking for? Sorry if that's a dumb question 🙃

**Reply by rrousselGit** [reply]
Is 25 providers rebuilding a number close to what you currently have? Because if so, that pretty much eliminates the fact that Riverpod may be the issue

You'd likely want to add a lot of stopwatchs around anything that is not an "await". In short, anything that could be blocking the main thread.

For example maybe you have a Json decoding of a large object. That could be an issue

**Reply by rrousselGit** [reply]
You'll likely have a simpler time debugging it using the CPU profiler instead of the jank frame tool too.

I'd suggest manipulating your providers by hand in a benchmark. You could create a ProviderContainer and run a few specific providers in a fixed scenario

This way you wouldn't be dependent on the concept of "frame". And the CPU profiler would be more tailored to your need

**Reply by btrautmann** [reply]
> You'd likely want to add a lot of stopwatchs around anything that is not an "await". In short, anything that could be blocking the main thread.

So the most common thing these `Provider`s do is iterating (most often through something like `fold`) through the transactions (which could be in the thousands of objects) and creating something like a `Map` of the transaction date to the transaction object.

I took a heavy-handed approach and moved these heavier functions (which I measured to sometimes take 200-400ms) to a background isolate via `compute`. This had the very positive/immediate effect of reducing `ProviderScope` build time to ~2ms. Unfortunately, due to (I assume) not reusing the same isolate, and these functions being called often, I am seeing really slow performance. There's no more jank, but the actual production of data itself is taking much longer. So I am (I think) going to need to find a way to reuse a single isolate assuming I'm correct that this is my issue.

Either way, I think we can firmly say Riverpod isn't the issue here, and I appreciate your help in getting to the bottom of this!
